livekit-plugins/livekit-plugins-assemblyai/livekit/plugins/assemblyai/stt.py

```diff
@@ -0,0 +1,384 @@
+        if self._closed:
+            raise ValueError("cannot push frame to closed stream")
+
+        self._queue.put_nowait(frame)
+
+    async def aclose(self, *, wait: bool = True) -> None:
+        self._closed = True
+        self._queue.put_nowait(SpeechStream._CLOSE_MSG)
+
+        if not wait:
+            self._main_task.cancel()
+
+        with suppress(asyncio.CancelledError):
+            await self._main_task
+
+        await self._session.close()
+
+    async def _main_task(self) -> None:
+        return self._run(self._max_retry)
+
+    async def _run(self, max_retry: int) -> None:
+        """
+        Run a single websocket connection to AssemblyAI and make sure to reconnect
+        when something went wrong.
+        """
+        try:
+            retry_count = 0
+            while not self._closed:
+                try:
+                    live_config = {
+                        "sample_rate": self._opts.sample_rate,
+                        "word_boost": self._opts.word_boost,
+                        "encoding": self._opts.encoding,
+                        "disable_partial_transcripts": self._opts.disable_partial_transcripts,
+                        "enable_extra_session_information": self._opts.enable_extra_session_information,
+                    }
+
+                    headers = {
+                        "Authorization": self._api_key,
+                        "Content-Type": "application/json",
+                    }
+
+                    ws_url = "wss://api.assemblyai.com/v2/realtime/ws"
+                    url = f"{ws_url}?{urlencode(live_config).lower()}"
+                    ws = await self._session.ws_connect(url, headers=headers)
+                    retry_count = 0  # connected successfully, reset the retry_count
+
+                    await self._run_ws(ws)
+                except Exception:
+                    # Something went wrong, retry the connection
+                    if retry_count >= max_retry:
+                        print(
+                            f"failed to connect to AssemblyAI after {max_retry} tries"
+                        )
+                        break
+
+                    retry_delay = min(retry_count * 2, 10)  # max 10s
+                    retry_count += 1  # increment after calculating the delay, the first retry should happen directly
+
+                    print(
+                        f"AssemblyAI connection failed, retrying in {retry_delay}s",
+                    )
+                    await asyncio.sleep(retry_delay)
+        except Exception:
+            print("AssemblyAI task failed")
+        finally:
+            self._event_queue.put_nowait(None)
+
+    async def _run_ws(self, ws: aiohttp.ClientWebSocketResponse) -> None:
+        """
+        This method can throw ws errors, these are handled inside the _run method
+        """
+
+        closing_ws = False
+        END_UTTERANCE_SILENCE_THRESHOLD_MSG = json.dumps(
```

Comment:
One note: on line 263 below you'll raise an error if the data going over the queue is neither an `rtc.AudioFrame` nor a `_CLOSE_MSG`, so sending this message (which is a string) causes the loop to always fail

Reply:
fixed, thanks!